CONTRIBUTING.md

```diff
@@ -1,208 +1,179 @@
+ - Before submitting changes, run the test suite to make sure that nothing is broken: `./gradlew check`
+
+**Submitting Changes**
+
+Please follow these steps in order to make sure your code has a chance to make it into the Elasticsearch codebase:
+
+1. Test your changes
+		 - Run the test suite to make sure nothing is broken. See the [TESTING](https://github.com/elastic/elasticsearch/blob/master/TESTING.asciidoc) file for help running tests
+2.  Sign the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/) (CLA)
+	- We are not asking you to assign copyright to us, but to give us the right to distribute your code without restriction
+	 - The CLA is used to assure our users of the origin and continuing existence of the code
+	- You only need to sign the CLA once
+3. Rebase your changes
+	- Update your local repository with the most recent code from the main Elasticsearch repository and rebase your branch on top of the latest master branch
+	- We prefer your initial changes to be squashed into a single commit. Later, if we ask you to make changes, add them as separate commits. This makes them easier to review.
+	- As a final step before merging, we will either ask you to squash all commits yourself or we will do it for you
 4. Submit a pull request
+	- Push your local changes to your forked copy of the repository and submit a [pull request](https://help.github.com/articles/using-pull-requests) 
+	- In the pull request, please follow these guidelines:
+		1. **Title**: Sum up the changes that you have made
+		2. **Body**: Provide more details about what your changes do and mention the number of the issue where discussion has taken place, eg "Closes #123."
+
+Once you have completed the above steps, sit back and wait. A discussion about the pull request will most likely be created. If any changes are needed, we would love to work with you to get your pull request merged into Elasticsearch!
+
+Please adhere to the general guideline that you should **never** force push to a **publicly shared** branch. Once you have opened your pull request, you should consider your branch publicly shared. Instead of force pushing, just add incremental commits. This is generally easier on your reviewers. If you need to pick up changes from master, you can merge master into your branch. A reviewer might ask you to rebase a long-running pull request in which case force pushing **is okay** for that request. Note that **squashing** at the end of the review process should also **not** be done. That can be done when the pull request is [integrated via GitHub](https://github.com/blog/2141-squash-your-commits).
+
+## Contributing as Part of a Class
+
+In general, Elasticsearch is happy to accept contributions that were created as part of a class. However, we strongly advise against making the contribution as part of the class. If you have written code for a class, feel free to submit it.
```

Comment:
I wrote this section originally and now that I reread it I'm not sure it makes much sense. But I'm not sure that your changes are better.

The gist is:
1. For the sake of all that is holy do not make someone's grade contingent on getting something merged to our repo.
2. If you feel like writing code for Elasticsearch as part of a class, we'll review it and merge it, but in our own time.
3. Do not tell a bunch of students to go and cookie lick issues and then make a half hearted attempt at solving them. I don't care how good you think your students are, we can't handle 30 new contributors in the span of a week. We certainly can't handle them if they aren't doing this of their own free will, excited to make changes and get them merged.

Most of this comes down to time and resources. I'd love new contributors and am fine with them being in university. But we don't operate on academic timelines. Sometimes we're faster and sometimes we're slower. And I'm really worried that the incentives for students to contribute are messed up. I've had folks say they only plan to open the PR but not to do the work suggested by the reviewers to merge it because that is all their class wanted from them....